Mathlib/Combinatorics/SetFamily/Shatter.lean

```diff
@@ -0,0 +1,196 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Data.Finset.Sort
+import Mathlib.Data.Nat.Interval
+import Mathlib.Order.UpperLower.Basic
+import Mathlib.Combinatorics.SetFamily.Compression.Down
+
+/-!
+# Shattering families
+
+This file defines the shattering property and VC-dimension of set families.
+
+## Main declarations
+
+* `Finset.Shatter`: The shattering property.
+* `Finset.shatterer`: The set family of sets shattered by a set family.
+* `Finset.vcDim`: The Vapnik-Chervonenkis dimension.
+
+## TODO
+
+* Order-shattering
+* Strong shattering
+-/
+
+open scoped BigOperators FinsetFamily
+
+namespace Finset
+variable {α : Type*} [DecidableEq α] {𝒜 ℬ : Finset (Finset α)} {s t : Finset α} {a : α} {n : ℕ}
+
+/-- A set family `𝒜` shatters a set `s` if all subsets of `s` can be obtained as the intersection
+of `s` and some element of the set family, and we denote this `𝒜.Shatter s`. We also say that `s` is
+traced by `𝒜`. -/
+def Shatter (𝒜 : Finset (Finset α)) (s : Finset α) : Prop := ∀ ⦃t⦄, t ⊆ s → ∃ u ∈ 𝒜, s ∩ u = t
```

Comment:
At risk of bikeshedding, should this property be called `Shatters`? `𝒜.Shatters s` looks cleaner to my eyes.